FlowCrypt/Functionality/Services/Key Services/PassPhraseStorage.swift

```diff
@@ -0,0 +1,116 @@
+//
+//  PassPhraseStorageService.swift
+//  FlowCrypt
+//
+//  Created by Anton Kharchevskyi on 02.06.2021.
+//  Copyright © 2021 FlowCrypt Limited. All rights reserved.
+//
+
+import UIKit
+
+protocol PassPhraseStorageType {
+    func getPassPhrases() -> [PassPhrase]
+    func savePassPhrase(with passPhrase: PassPhrase, inStorage: Bool)
+    func updatePassPhrase(with passPhrase: PassPhrase, inStorage: Bool)
+}
+
+protocol EmailProviderType {
+    var email: String? { get }
+}
+
+final class PassPhraseStorage: PassPhraseStorageType {
+    private lazy var logger = Logger.nested(Self.self)
+
+    let currentUserEmail: String?
+    let storage: EncryptedPassPhraseStorage
+    let localStorage: LocalPassPhraseStorageType
+    let timeoutContext: (component: Calendar.Component, timeout: Int)
+
+    /// used for tests only, otherwise seconds will be used
+    let isHours: Bool
+
+    init(
+        storage: EncryptedPassPhraseStorage,
+        localStorage: LocalPassPhraseStorageType = LocalPassPhraseStorage.shared,
+        timeoutContext: (Calendar.Component, Int) = (.hour, 4),
+        emailProvider: EmailProviderType,
+        isHours: Bool = true
+    ) {
+        self.storage = storage
+        self.localStorage = localStorage
+        self.timeoutContext = timeoutContext
+        self.currentUserEmail = emailProvider.email
+        self.isHours = isHours
+    }
+
+    func savePassPhrase(with passPhrase: PassPhrase, inStorage: Bool) {
+        if inStorage {
+            logger.logInfo("Save to storage \(passPhrase.longid)")
+            storage.addPassPhrase(object: PassPhraseObject(passPhrase))
+        } else {
+            logger.logInfo("Save locally \(passPhrase.longid)")
+
+            let locallPassPhrase = LocalPassPhrase(passPhrase: passPhrase, date: Date())
+            localStorage.save(passPhrase: locallPassPhrase)
+
+            let alreadySaved = storage.getPassPhrases()
+
+            if alreadySaved.contains(where: { $0.longid == passPhrase.longid }) {
+                storage.removePassPhrase(object: PassPhraseObject(passPhrase))
+            }
+        }
+    }
+
+    func updatePassPhrase(with passPhrase: PassPhrase, inStorage: Bool) {
+        if inStorage {
+            storage.updatePassPhrase(object: PassPhraseObject(passPhrase))
+        } else {
+            let updated = LocalPassPhrase(passPhrase: passPhrase, date: Date())
+            localStorage.save(passPhrase: updated)
+        }
+    }
+
+    func getPassPhrases() -> [PassPhrase] {
```

Comment:
I don't think this implements storage in memory at all - it stores pass phrases in the database, and invalidates them on retrieval if they are too old. This is not what I wanted, unless I overlooked something.

Reply:
I've made another issue for this so that we can get this PR merged, and then carefully evaluate/review the proper implementation later.

Reply:
But why?
```
func savePassPhrase(with passPhrase: PassPhrase, inStorage: Bool) {
        if inStorage {
            ...
        } else {
            ...
            localStorage.save(passPhrase: locallPassPhrase)
        }
    }
 ```
So if it's not in storage - then it ask localStorage to save it. Local storage is a shared instance(live only during lifetime of application), it doesn't have any persisted storages inside or any else. It just have one in memory `private(set) var passPhrases: Set<LocalPassPhrase> = []`

Reply:
I see - I'll take another look.

Reply:
You're right, sorry. I made another PR comment to rename things to be less confusing.